README.md
# lumos

![build](https://github.com/ximingwang/lumos/actions/workflows/github-ci.yaml/badge.svg)
[![codecov](https://codecov.io/gh/ximingwang/lumos/branch/develop/graph/badge.svg?token=6WJJOOMD2F)](https://codecov.io/gh/ximingwang/lumos)
![license](https://img.shields.io/github/license/ximingwang/lumos)
![Lumos](./assets/lumos.jpg)

> Lumos is still under active development and considered to be a work in progress.

Lumos is a full featured JavaScript/TypeScript based dapp framework for Nervos CKB. It aims to simplify dapp development on Nervos CKB. The [name](https://harrypotter.fandom.com/wiki/Lumos_Maxima) explains what the project is about: the wonderland of CKB, though vast and fertile, still remains dark in most of the parts, lumos tries to shed light upon the land, guiding people into this beautiful new world.

As of now, lumos contains the following components:

- [ckb-indexer](./packages/ckb-indexer): an cell/tx indexer base on [ckb-indexer](https://github.com/ximingwang/ckb-indexer).
- [BI](./packages/bi): a big number library for lumos.
- [toolkit](./packages/toolkit): JavaScript toolkit for Nervos CKB. It contains a series of independent tools that can aid develoment of CKB dapps. This is different from a normal CKB SDK, since it tries to minimize the work done in an SDK, while providing more utlities that could be handy.
- [lumos](./packages/lumos): A root package to integrate of common functions from the Lumos sub-package.
- [base](./packages/base): a base package containing common types and utilities that are used by most packages. If there is a CKB specific task you need to perform, you might want to look here first. Chances are they are already provided.
- [helpers](./packages/helpers): a helper package containing more utilities. The difference between `helpers` and `base`, is that `base` contains pure stateless functions, while `helpers` works in a more intrinsic way: it requires `config-manager` mentioned below to be setup.
- [common-scripts](./packages/common-scripts): integrations for known scripts on CKB. While we try our best to provide integrations for popular CKB scripts, people might be working on innovations everyday. As a result, we are also designing a set of APIs, so developers can freely integrate their own scripts into lumos for everyone to use. One integrated, `common-scripts` should be able to leverage those new scripts as well.
- [config-manager](./packages/config-manager): a manager for dealing with differences between different chains, such as mainnet, testnet, or numerous dev chains. We abstract each chain into individual config file. Once loaded, config manager will be able to handle the chain specific logic, so you don't have to deal with this in your own code.
- [transaction-manager](./packages/transaction-manager): a transaction manager for CKB. One problem with UTXO based blockchains, is that a certain amount of gap period exists between a transaction is accepted by a blockchain, and when it is actually committed on chain. During this gap, new cells created by the pending transaction will not be available. Transaction manager package takes care of this. It wraps an indexer instance, and makes sure cells created in pending transactions, are also exposed and available for assembling new transactions. This means you are no longer bounded to one transaction at a time, you can freely send series of transactions as you wish.
- [hd](./packages/hd): an HD wallet manager for CKB. It support mnemonic and keystore, compatible with `Neuron` and `ckb-cli`, you can load keystore from `Neuron` or `ckb-cli` directly and import mnemonic generated by `Neuron`.
- [hd-cache](./packages/hd-cache): an HD cache manager for CKB. It build a memory cache for derived addresses and live cells of these addresses.
- [rpc](./packages/rpc): RPC module for CKB RPC. Provide type definations for CKB RPC interface.

## Examples

we have provided some use cases for Lumos, such as interactions with MetaMask, transfers CKB, address conversions, etc. which you can find in [examples](./examples)

## Building

### Requirements

- [Node.js](https://nodejs.org)
- [Yarn](https://yarnpkg.com/)
- [node-gyp](https://github.com/nodejs/node-gyp)

```bash
sudo apt-get update
sudo apt install nodejs
npm install --global yarn
sudo apt install build-essential
```

### Build

```bash
yarn run build
```

### Test (ava)

```bash
yarn run test
```

### Test Coverage (c8)

```bash
yarn run test-coverage
```

### Format & Lint

```bash
yarn run lint
```

### Clean

```bash
yarn run clean
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.18.0-rc5.1](https://github.com/nervosnetwork/lumos/compare/v0.18.0-rc5...v0.18.0-rc5.1) (2022-02-24)


### Bug Fixes

* **examples:** add @ckb-lumos/lumos dependence to pw-lock and secp256k1 ([55ddc16](https://github.com/nervosnetwork/lumos/commit/55ddc16ec9446bf080dfeedc2f55b04713deac1c))
* **examples:** add @ckb-lumos/lumos dependence which is missing ([8aedfdb](https://github.com/nervosnetwork/lumos/commit/8aedfdbd673cfdd5d1bff68fed97b448f560b30f))
* **hd-cache:** use ckb-lumos/base indexer type instand ckb-lumos/ckb-indexer indexer value ([4fd1bfd](https://github.com/nervosnetwork/lumos/commit/4fd1bfdf79006527e6599dea0020c9a2d8cd770a))


### Features

* **common-scripts:** added P2PKH general function ([#254](https://github.com/nervosnetwork/lumos/issues/254)) ([279f793](https://github.com/nervosnetwork/lumos/commit/279f7937f4da6cae7473b30c6b3f929b90e164e5))
* remove example ([145ffcb](https://github.com/nervosnetwork/lumos/commit/145ffcb2d9ea0a8f36b5b15214a75a7d42404bb4))
* remove indexer ([7b0cf5c](https://github.com/nervosnetwork/lumos/commit/7b0cf5c1c4993caabf9e05baf49a8cb0b4a4697b))
* **transaction-manager:** change nativeIndex.getLiveCellDetail to rpc.getLiveCell because nativeInder is not use any more ([086002a](https://github.com/nervosnetwork/lumos/commit/086002ac90758b09c6a83de00500fde8c5c9b136))





# [0.18.0-rc5.0](https://github.com/nervosnetwork/lumos/compare/v0.18.0-rc5...v0.18.0-rc5.0) (2022-02-24)


### Bug Fixes

* **examples:** add @ckb-lumos/lumos dependence to pw-lock and secp256k1 ([55ddc16](https://github.com/nervosnetwork/lumos/commit/55ddc16ec9446bf080dfeedc2f55b04713deac1c))
* **examples:** add @ckb-lumos/lumos dependence which is missing ([8aedfdb](https://github.com/nervosnetwork/lumos/commit/8aedfdbd673cfdd5d1bff68fed97b448f560b30f))
* **hd-cache:** use ckb-lumos/base indexer type instand ckb-lumos/ckb-indexer indexer value ([4fd1bfd](https://github.com/nervosnetwork/lumos/commit/4fd1bfdf79006527e6599dea0020c9a2d8cd770a))


### Features

* **common-scripts:** added P2PKH general function ([#254](https://github.com/nervosnetwork/lumos/issues/254)) ([279f793](https://github.com/nervosnetwork/lumos/commit/279f7937f4da6cae7473b30c6b3f929b90e164e5))
* **transaction-manager:** change nativeIndex.getLiveCellDetail to rpc.getLiveCell because nativeInder is not use any more ([086002a](https://github.com/nervosnetwork/lumos/commit/086002ac90758b09c6a83de00500fde8c5c9b136))


examples/ckb-indexer-collector.ts
// This example shows how to use the CkbIndexer to collect cells

import { Script, Indexer as CkbIndexer, helpers, config } from "@ximingwang/lumos";

config.initializeConfig(config.predefined.AGGRON4);

const CKB_RPC_URL = "https://testnet.ckb.dev/rpc";
const CKB_INDEXER_URL = "https://testnet.ckb.dev/indexer";
const indexer = new CkbIndexer(CKB_INDEXER_URL, CKB_RPC_URL);

async function capacityOf(lock: Script): Promise<bigint> {
  const collector = indexer.collector({ lock });

  let balance = 0n;
  for await (const cell of collector.collect()) {
    balance += BigInt(cell.cell_output.capacity);
  }

  return balance;
}

async function main() {
  const address = "ckt1qyqxgftlqzmtv05cwcyl4xlz6ryx6dgsyrasjrp27t";
  const lock: Script = helpers.parseAddress(address);

  const balance = await capacityOf(lock);

  const integer = balance / 10n ** 8n;
  const fraction = balance % 10n ** 8n;

  console.log(`lock of ${address} is`, lock, "\n");
  console.log(`total CKB of ${address} is ${integer + "." + fraction}`);
}

main();


examples/secp256k1-transfer/lib.ts
import {
  Indexer,
  helpers,
  Address,
  Script,
  RPC,
  hd,
  config,
  Cell,
  commons,
  core,
  WitnessArgs,
  toolkit,
} from "@ximingwang/lumos";
import { values } from "@ximingwang/base";
const { ScriptValue } = values;

export const { AGGRON4 } = config.predefined;

const CKB_RPC_URL = "https://testnet.ckb.dev/rpc";
const CKB_INDEXER_URL = "https://testnet.ckb.dev/indexer";
const rpc = new RPC(CKB_RPC_URL);
const indexer = new Indexer(CKB_INDEXER_URL, CKB_RPC_URL);

type Account = {
  lockScript: Script;
  address: Address;
  pubKey: string;
};
export const generateAccountFromPrivateKey = (privKey: string): Account => {
  const pubKey = hd.key.privateToPublic(privKey);
  const args = hd.key.publicKeyToBlake160(pubKey);
  const template = AGGRON4.SCRIPTS["SECP256K1_BLAKE160"]!;
  const lockScript = {
    code_hash: template.CODE_HASH,
    hash_type: template.HASH_TYPE,
    args: args,
  };
  const address = helpers.generateAddress(lockScript, { config: AGGRON4 });
  return {
    lockScript,
    address,
    pubKey,
  };
};

export async function capacityOf(address: string): Promise<bigint> {
  const collector = indexer.collector({
    lock: helpers.parseAddress(address, { config: AGGRON4 }),
  });

  let balance = 0n;
  for await (const cell of collector.collect()) {
    balance += BigInt(cell.cell_output.capacity);
  }

  return balance;
}

interface Options {
  from: string;
  to: string;
  amount: string;
  privKey: string;
}

export async function transfer(options: Options): Promise<string> {
  let txSkeleton = helpers.TransactionSkeleton({});
  const fromScript = helpers.parseAddress(options.from, { config: AGGRON4 });
  const toScript = helpers.parseAddress(options.to, { config: AGGRON4 });

  // additional 0.001 ckb for tx fee
  // the tx fee could calculated by tx size
  // this is just a simple example
  const neededCapacity = BigInt(options.amount) + 100000n;
  let collectedSum = 0n;
  const collected: Cell[] = [];
  const collector = indexer.collector({ lock: fromScript, type: "empty" });
  for await (const cell of collector.collect()) {
    collectedSum += BigInt(cell.cell_output.capacity);
    collected.push(cell);
    if (collectedSum >= neededCapacity) break;
  }

  if (collectedSum < neededCapacity) {
    throw new Error("Not enough CKB");
  }

  const transferOutput: Cell = {
    cell_output: {
      capacity: "0x" + BigInt(options.amount).toString(16),
      lock: toScript,
    },
    data: "0x",
  };

  const changeOutput: Cell = {
    cell_output: {
      capacity: "0x" + BigInt(collectedSum - neededCapacity).toString(16),
      lock: fromScript,
    },
    data: "0x",
  };

  txSkeleton = txSkeleton.update("inputs", (inputs) =>
    inputs.push(...collected)
  );
  txSkeleton = txSkeleton.update("outputs", (outputs) =>
    outputs.push(transferOutput, changeOutput)
  );
  txSkeleton = txSkeleton.update("cellDeps", (cellDeps) =>
    cellDeps.push({
      out_point: {
        tx_hash: AGGRON4.SCRIPTS.SECP256K1_BLAKE160.TX_HASH,
        index: AGGRON4.SCRIPTS.SECP256K1_BLAKE160.INDEX,
      },
      dep_type: AGGRON4.SCRIPTS.SECP256K1_BLAKE160.DEP_TYPE,
    })
  );

  const firstIndex = txSkeleton
    .get("inputs")
    .findIndex((input) =>
      new ScriptValue(input.cell_output.lock, { validate: false }).equals(
        new ScriptValue(fromScript, { validate: false })
      )
    );
  if (firstIndex !== -1) {
    while (firstIndex >= txSkeleton.get("witnesses").size) {
      txSkeleton = txSkeleton.update("witnesses", (witnesses) =>
        witnesses.push("0x")
      );
    }
    let witness: string = txSkeleton.get("witnesses").get(firstIndex)!;
    const newWitnessArgs: WitnessArgs = {
      /* 65-byte zeros in hex */
      lock:
        "0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    };
    if (witness !== "0x") {
      const witnessArgs = new core.WitnessArgs(new toolkit.Reader(witness));
      const lock = witnessArgs.getLock();
      if (
        lock.hasValue() &&
        new toolkit.Reader(lock.value().raw()).serializeJson() !==
          newWitnessArgs.lock
      ) {
        throw new Error(
          "Lock field in first witness is set aside for signature!"
        );
      }
      const inputType = witnessArgs.getInputType();
      if (inputType.hasValue()) {
        newWitnessArgs.input_type = new toolkit.Reader(
          inputType.value().raw()
        ).serializeJson();
      }
      const outputType = witnessArgs.getOutputType();
      if (outputType.hasValue()) {
        newWitnessArgs.output_type = new toolkit.Reader(
          outputType.value().raw()
        ).serializeJson();
      }
    }
    witness = new toolkit.Reader(
      core.SerializeWitnessArgs(
        toolkit.normalizers.NormalizeWitnessArgs(newWitnessArgs)
      )
    ).serializeJson();
    txSkeleton = txSkeleton.update("witnesses", (witnesses) =>
      witnesses.set(firstIndex, witness)
    );
  }

  txSkeleton = commons.common.prepareSigningEntries(txSkeleton);
  const message = txSkeleton.get("signingEntries").get(0)?.message;
  const Sig = hd.key.signRecoverable(message!, options.privKey);
  const tx = helpers.sealTransaction(txSkeleton, [Sig]);
  const hash = await rpc.send_transaction(tx, "passthrough");
  console.log("The transaction hash is", hash);

  return hash;
}


examples/secp256k1-multisig-transfer/index.ts
import {
  Indexer,
  helpers,
  Script,
  RPC,
  hd,
  config,
  Cell,
  commons,
  core,
  WitnessArgs,
  toolkit,
} from "@ximingwang/lumos";
import { values } from "@ximingwang/base";
import { parseFromInfo, MultisigScript } from "@ximingwang/common-scripts/lib/from_info";
const { ScriptValue } = values;

const { AGGRON4 } = config.predefined;

const CKB_RPC_URL = "https://testnet.ckb.dev/rpc";
const CKB_INDEXER_URL = "https://testnet.ckb.dev/indexer";
const rpc = new RPC(CKB_RPC_URL);
const indexer = new Indexer(CKB_INDEXER_URL, CKB_RPC_URL);

const ALICE = {
  PRIVATE_KEY: "0x2c56a92a03d767542222432e4f2a0584f01e516311f705041d86b1af7573751f",
  ARGS: "0x3d35d87fac0008ba5b12ee1c599b102fc8f5fdf8",
};

const BOB = {
  PRIVATE_KEY: "0x3bc65932a75f76c5b6a04660e4d0b85c2d9b5114efa78e6e5cf7ad0588ca09c8",
  ARGS: "0x99dbe610c43186696e1f88cb7b59252d4c92afda",
};

const CHARLES = {
  PRIVATE_KEY: "0xbe06025fbd8c74f65a513a28e62ac56f3227fcb307307a0f2a0ef34d4a66e81f",
  ARGS: "0xc055df68fdd47c6a5965b9ab21cd6825d8696a76",
};

const TO_ADDRESS = "ckt1qyqptxys5l9vk39ft0hswscxgseawc77y2wqlr558h";

/**
 * Generate fromInfo for multisig transfer.
 * @param R The provided signatures must match at least the first R items of the Pubkey list.
 * @param M M of N signatures must be provided to unlock the cell. N equals to the size of publicKeyHashes.
 * @param publicKeyHashes The list of Lock Args generated by the blake160 function that extracts
 * the first 20 bytes of a public key hash.
 *
 * R, M are single byte unsigned integers that ranges from 0 to 255.
 * R must no more than M.
 */
function generateMofNMultisigInfo(R: number, M: number, publicKeyHashes: string[]): MultisigScript {
  return {
    R,
    M,
    publicKeyHashes,
  };
}

type Account = {
  fromScript: Script;
  multisigScript: string | undefined;
};

function generateAccountFromMultisigInfo(fromInfo: MultisigScript): Account {
  const { fromScript, multisigScript } = parseFromInfo(fromInfo, {
    config: AGGRON4,
  });
  return {
    fromScript,
    multisigScript,
  };
}

interface Options {
  fromInfo: MultisigScript;
  toAddress: string;
  amount: bigint;
  privKeys: string[];
}

export async function transfer(options: Options): Promise<string> {
  let txSkeleton = helpers.TransactionSkeleton({});

  const { fromScript, multisigScript } = generateAccountFromMultisigInfo(options.fromInfo);
  const toScript = helpers.parseAddress(options.toAddress, { config: AGGRON4 });

  // additional 0.001 ckb for tx fee
  // the tx fee could calculated by tx size
  // this is just a simple example
  const neededCapacity = BigInt(options.amount) + 100000n;
  let collectedSum = 0n;
  const collected: Cell[] = [];
  const collector = indexer.collector({ lock: fromScript, type: "empty" });
  for await (const cell of collector.collect()) {
    collectedSum += BigInt(cell.cell_output.capacity);
    collected.push(cell);
    if (collectedSum >= neededCapacity) break;
  }

  if (collectedSum < neededCapacity) {
    throw new Error("Not enough CKB");
  }

  const transferOutput: Cell = {
    cell_output: {
      capacity: "0x" + BigInt(options.amount).toString(16),
      lock: toScript,
    },
    data: "0x",
  };

  const changeOutput: Cell = {
    cell_output: {
      capacity: "0x" + BigInt(collectedSum - neededCapacity).toString(16),
      lock: fromScript,
    },
    data: "0x",
  };

  txSkeleton = txSkeleton.update("inputs", (inputs) => inputs.push(...collected));
  txSkeleton = txSkeleton.update("outputs", (outputs) => outputs.push(transferOutput, changeOutput));
  txSkeleton = txSkeleton.update("cellDeps", (cellDeps) =>
    cellDeps.push({
      out_point: {
        tx_hash: AGGRON4.SCRIPTS.SECP256K1_BLAKE160_MULTISIG.TX_HASH,
        index: AGGRON4.SCRIPTS.SECP256K1_BLAKE160_MULTISIG.INDEX,
      },
      dep_type: AGGRON4.SCRIPTS.SECP256K1_BLAKE160_MULTISIG.DEP_TYPE,
    })
  );

  const firstIndex = txSkeleton
    .get("inputs")
    .findIndex((input) =>
      new ScriptValue(input.cell_output.lock, { validate: false }).equals(
        new ScriptValue(fromScript, { validate: false })
      )
    );
  if (firstIndex !== -1) {
    while (firstIndex >= txSkeleton.get("witnesses").size) {
      txSkeleton = txSkeleton.update("witnesses", (witnesses) => witnesses.push("0x"));
    }
    let witness: string = txSkeleton.get("witnesses").get(firstIndex)!;
    let newWitnessArgs: WitnessArgs;
    const SECP_SIGNATURE_PLACEHOLDER =
      "0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000";

    newWitnessArgs = {
      lock: "0x" + multisigScript!.slice(2) + SECP_SIGNATURE_PLACEHOLDER.slice(2).repeat(options.fromInfo.M),
    };

    if (witness !== "0x") {
      const witnessArgs = new core.WitnessArgs(new toolkit.Reader(witness));
      const lock = witnessArgs.getLock();
      if (lock.hasValue() && new toolkit.Reader(lock.value().raw()).serializeJson() !== newWitnessArgs.lock) {
        throw new Error("Lock field in first witness is set aside for signature!");
      }
      const inputType = witnessArgs.getInputType();
      if (inputType.hasValue()) {
        newWitnessArgs.input_type = new toolkit.Reader(inputType.value().raw()).serializeJson();
      }
      const outputType = witnessArgs.getOutputType();
      if (outputType.hasValue()) {
        newWitnessArgs.output_type = new toolkit.Reader(outputType.value().raw()).serializeJson();
      }
    }
    witness = new toolkit.Reader(
      core.SerializeWitnessArgs(toolkit.normalizers.NormalizeWitnessArgs(newWitnessArgs))
    ).serializeJson();
    txSkeleton = txSkeleton.update("witnesses", (witnesses) => witnesses.set(firstIndex, witness));
  }

  txSkeleton = commons.common.prepareSigningEntries(txSkeleton);
  const message = txSkeleton.get("signingEntries").get(0)?.message;

  let pubkeyHashN: string = "";
  options.fromInfo.publicKeyHashes.forEach((publicKeyHash) => {
    pubkeyHashN += publicKeyHash.slice(2);
  });

  let sigs: string = "";
  options.privKeys.forEach((privKey) => {
    if (privKey !== "") {
      let sig = hd.key.signRecoverable(message!, privKey);
      sig = sig.slice(2);
      sigs += sig;
    }
  });

  sigs =
    "0x00" +
    ("00" + options.fromInfo.R.toString(16)).slice(-2) +
    ("00" + options.fromInfo.M.toString(16)).slice(-2) +
    ("00" + options.fromInfo.publicKeyHashes.length.toString(16)).slice(-2) +
    pubkeyHashN +
    sigs;

  const tx = helpers.sealTransaction(txSkeleton, [sigs]);
  const hash = await rpc.send_transaction(tx, "passthrough");
  console.log("The transaction hash is", hash);

  return hash;
}

// Multisig transfer example
function main() {
  const fromInfo = generateMofNMultisigInfo(2, 2, [ALICE.ARGS, BOB.ARGS, CHARLES.ARGS]);
  const privKeys = [ALICE.PRIVATE_KEY, BOB.PRIVATE_KEY];
  transfer({ fromInfo, toAddress: TO_ADDRESS, amount: 8800000000n, privKeys });
}
main();
